File: examples/proxy.rs
//! Example: simple L4 proxy

use std::{env, io};

use tokio::net::{TcpListener, TcpStream};

// #[tokio::main(flavor = "current_thread")]
#[tokio::main]
async fn main() -> io::Result<()> {
    println!("PID is {}", std::process::id());

    use tracing::level_filters::LevelFilter;
    use tracing_subscriber::layer::SubscriberExt;
    use tracing_subscriber::util::SubscriberInitExt;
    use tracing_subscriber::{EnvFilter, Layer};

    let (w, _g) = tracing_appender::non_blocking(io::stdout());
    let fmt_layer = tracing_subscriber::fmt::layer().with_writer(w).with_filter(
        EnvFilter::builder()
            .with_default_directive(LevelFilter::DEBUG.into())
            .from_env_lossy()
            .add_directive("otel::tracing=trace".parse().unwrap())
            .add_directive("h2=error".parse().unwrap())
            .add_directive("tower=error".parse().unwrap())
            .add_directive("hyper=error".parse().unwrap()),
    );

    tracing_subscriber::registry()
        .with(fmt_layer)
        // .with(console_subscriber::spawn())
        .init();

    tokio::select! {
        res = serve() => {
            if let Err(err) = res {
                eprintln!("Serve failed {err}");
            }
        }
        _ = tokio::signal::ctrl_c() => {
            println!("Received Ctrl + C, shutting down");
        }
    }

    Ok(())
}

async fn serve() -> io::Result<()> {
    let listener = TcpListener::bind(
        env::var("EXAMPLE_LISTEN_ADDR").unwrap_or_else(|_| "0.0.0.0:5201".to_string()),
    )
    .await?;

    loop {
        let (incoming, remote_addr) = match listener.accept().await {
            Ok(accepted) => accepted,
            Err(e) if e.kind() == io::ErrorKind::ConnectionAborted => {
                println!("Connection aborted.");
                continue;
            }
            Err(e) => {
                eprintln!("Failed to accept: {e:#?}");
                break Err(e);
            }
        };

        println!("Process incoming connection from {remote_addr}");

        tokio::spawn(forwarding(incoming));
    }
}

async fn forwarding(mut stream1: TcpStream) -> io::Result<()> {
    let stream2 = TcpStream::connect(
        env::var("EXAMPLE_REMOTE_ADDR").unwrap_or_else(|_| "127.0.0.1:5202".to_string()),
    )
    .await;

    let mut stream2 = match stream2 {
        Ok(s) => s,
        Err(e) => {
            eprintln!("Failed to connect to remote server: {e}");
            return Err(e);
        }
    };

    let instant = std::time::Instant::now();

    // let result = tokio_splice::zero_copy_bidirectional(&mut stream1, &mut
    // stream2).await;

    // let limit = env::var("LIMIT")
    //     .ok()
    //     .and_then(|r| r.parse().ok())
    //     .unwrap_or(1u64 * 1000 * 1000);

    // println!("Rate limit is set to {limit} B/s");

    // let limit = RateLimit::new(NonZeroU64::new(limit).unwrap());

    // let rate_limiter_clone = rate_limiter.clone();
    // tokio::spawn(async move {
    //     loop {
    //         tokio::time::sleep(Duration::from_secs(10)).await;
    //         rate_limiter_clone.set_total(NonZeroU64::new(100 * 1024).unwrap());
    //         println!("Rate limiter updated to 100 KiB/s");

    //         tokio::time::sleep(Duration::from_secs(10)).await;
    //         rate_limiter_clone.set_total(NonZeroU64::new(8 * 1024).unwrap());
    //         println!("Rate limiter updated to 8 KiB/s");

    //         tokio::time::sleep(Duration::from_secs(10)).await;
    //         rate_limiter_clone.set_total(NonZeroU64::new(8 * 1024 *
    // 1024).unwrap());         println!("Rate limiter updated to 8 MiB/s");

    //         tokio::time::sleep(Duration::from_secs(10)).await;
    //         rate_limiter_clone.set_total(NonZeroU64::new(300 * 1024).unwrap());
    //         println!("Rate limiter updated to 300 KiB/s");

    //         // tokio::time::sleep(Duration::from_secs(10)).await;
    //         // rate_limiter_clone.set_total(NonZeroU64::new(16 * 1000 *
    //         // 1000).unwrap()); println!("Rate limiter updated to 16
    //         // MB/s");

    //         // tokio::time::sleep(Duration::from_secs(10)).await;

    //         // rate_limiter_clone.set_total(NonZeroU64::new(500 *
    //         // 1000).unwrap()); println!("Rate limiter updated to
    //         // 500 KB/s");

    //         // tokio::time::sleep(Duration::from_secs(10)).await;
    //         // rate_limiter_clone.set_disable();
    //         // println!("Rate limiter updated to unlimited");
    //     }
    // });

    let io_sl2sr = tokio_splice2::context::SpliceIoCtx::prepare()?.into_io();
    let io_sr2sl = tokio_splice2::context::SpliceIoCtx::prepare()?.into_io();

    let traffic = tokio_splice2::io::SpliceBidiIo { io_sl2sr, io_sr2sl }
        .execute(&mut stream1, &mut stream2)
        .await;
    // let result = realm_io::bidi_zero_copy(&mut stream1, &mut stream2).await;

    let total = traffic.sum();
    // let total = traffic.0 + traffic.1;
    let cost = instant.elapsed();
    println!(
        "Forwarded traffic: total: {total} B, time: {:.2} s, avg: {:.4} B/s, error: {:?}",
        cost.as_secs_f64(),
        total as f64 / cost.as_secs_f64(),
        traffic.error
    );

    Ok(())
}

File: examples/proxy.go
package main

import (
	"context"
	"fmt"
	"io"
	"log"
	"net"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"
)

func main() {
	fmt.Printf("PID is %d\n", os.Getpid())

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		if err := serve(ctx); err != nil {
			log.Printf("Serve failed: %v", err)
		}
	}()

	<-sigChan
	fmt.Println("Received Ctrl + C, shutting down")
	cancel()

	time.Sleep(100 * time.Millisecond)
}

func serve(ctx context.Context) error {
	listenAddr := os.Getenv("EXAMPLE_LISTEN_ADDR")
	if listenAddr == "" {
		listenAddr = "0.0.0.0:5201"
	}

	listener, err := net.Listen("tcp", listenAddr)
	if err != nil {
		return fmt.Errorf("failed to listen on %s: %w", listenAddr, err)
	}
	defer listener.Close()

	fmt.Printf("Listening on %s\n", listenAddr)

	for {
		select {
		case <-ctx.Done():
			return nil
		default:
		}

		if tcpListener, ok := listener.(*net.TCPListener); ok {
			tcpListener.SetDeadline(time.Now().Add(100 * time.Millisecond))
		}

		conn, err := listener.Accept()
		if err != nil {
			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
				continue
			}
			if netErr, ok := err.(net.Error); ok && netErr.Temporary() {
				log.Printf("Temporary accept error: %v", err)
				continue
			}
			return fmt.Errorf("failed to accept: %w", err)
		}

		remoteAddr := conn.RemoteAddr()
		fmt.Printf("Process incoming connection from %s\n", remoteAddr)

		go forwarding(conn)
	}
}

func forwarding(stream1 net.Conn) error {
	defer stream1.Close()

	remoteAddr := os.Getenv("EXAMPLE_REMOTE_ADDR")
	if remoteAddr == "" {
		remoteAddr = "127.0.0.1:5202"
	}

	stream2, err := net.Dial("tcp", remoteAddr)
	if err != nil {
		log.Printf("Failed to connect to remote server: %v", err)
		return err
	}
	defer stream2.Close()

	startTime := time.Now()

	result, err := copyBidirectional(stream1, stream2)

	elapsed := time.Since(startTime)

	fmt.Printf("Forwarded traffic: %+v, avg: %.4f B/s\n", result, float64((result.BytesForward+result.BytesReverse))/elapsed.Seconds())

	if err != nil {
		log.Printf("Failed to copy data: %v", err)
		return err
	}

	return nil
}

type TrafficStats struct {
	BytesForward uint64
	BytesReverse uint64
}

func (t TrafficStats) String() string {
	return fmt.Sprintf("TrafficStats { bytes_forward: %d, bytes_reverse: %d }",
		t.BytesForward, t.BytesReverse)
}

func copyBidirectional(conn1, conn2 net.Conn) (*TrafficStats, error) {
	var stats TrafficStats
	var wg sync.WaitGroup
	var err1, err2 error

	wg.Add(2)

	go func() {
		defer wg.Done()
		defer conn2.Close()
		n, err := io.Copy(conn2, conn1)
		stats.BytesForward = uint64(n)
		err1 = err
	}()

	go func() {
		defer wg.Done()
		defer conn1.Close()
		n, err := io.Copy(conn1, conn2)
		stats.BytesReverse = uint64(n)
		err2 = err
	}()

	wg.Wait()

	if err1 != nil && err1 != io.EOF {
		return &stats, err1
	}
	if err2 != nil && err2 != io.EOF {
		return &stats, err2
	}

	return &stats, nil
}
